drivers/sensor/ite_tach_it8xxx2/tach_ite_it8xxx2.c

```diff
@@ -0,0 +1,279 @@
+	/* Captured counts of tachometer */
+	uint32_t capture;
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev)	((const struct tach_it8xxx2_config *)(dev)->config)
+#define DRV_DATA(dev)	((struct tach_it8xxx2_data *)(dev)->data)
+
+static int tach_ch_valid(const struct device *dev, int tach_ch)
+{
+	const struct tach_it8xxx2_config *const config = DRV_CONFIG(dev);
+	volatile uint8_t *reg_tswctlr = (uint8_t *)config->reg_tswctlr;
+	int valid = 0;
+
+	switch (tach_ch) {
+	case TACH0_CHANNEL_A:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T0_MASK) == IT8XXX2_PWM_T0DVS)
+			valid = 1;
+		break;
+	case TACH0_CHANNEL_B:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T0_MASK) == IT8XXX2_PWM_T0_MASK)
+			valid = 1;
+		break;
+	case TACH1_CHANNEL_A:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T1_MASK) == IT8XXX2_PWM_T1DVS)
+			valid = 1;
+		break;
+	case TACH1_CHANNEL_B:
+		if ((*reg_tswctlr & IT8XXX2_PWM_T1_MASK) == IT8XXX2_PWM_T1_MASK)
+			valid = 1;
+		break;
+	default:
+		break;
+	}
+
+	return valid;
+}
+
+int tach_it8xxx2_sample_fetch(const struct device *dev,
+			      enum sensor_channel chan)
+{
+	const struct tach_it8xxx2_config *const config = DRV_CONFIG(dev);
+	volatile uint8_t *reg_fxtlrr = (uint8_t *)config->reg_fxtlrr;
+	volatile uint8_t *reg_fxtmrr = (uint8_t *)config->reg_fxtmrr;
+	volatile uint8_t *reg_tswctlr = (uint8_t *)config->reg_tswctlr;
+	int tach_ch = config->channel;
+	struct tach_it8xxx2_data *const data = DRV_DATA(dev);
+
+	if (chan != SENSOR_CHAN_RPM) {
+		LOG_ERR("Sensor chan %d, only support SENSOR_CHAN_RPM", chan);
+		return -ENOTSUP;
+	}
+
+	if (tach_ch_valid(dev, tach_ch)) {
+		/* If channel data of tachometer is valid, then save it */
+		data->capture = ((*reg_fxtmrr) << 8) | (*reg_fxtlrr);
+
+		if ((tach_ch == TACH0_CHANNEL_A) || (tach_ch == TACH0_CHANNEL_B))
+			/* Clear tachometer0 data valid status */
+			*reg_tswctlr |= IT8XXX2_PWM_T0DVS;
+		else
+			/* Clear tachometer1 data valid status */
+			*reg_tswctlr |= IT8XXX2_PWM_T1DVS;
+	} else {
+		/* If channel data of tachometer isn't valid, then clear it */
+		data->capture = 0;
+	}
+
+	return 0;
+}
+
+static int tach_it8xxx2_channel_get(const struct device *dev,
+				    enum sensor_channel chan,
+				    struct sensor_value *val)
+{
+	const struct tach_it8xxx2_config *const config = DRV_CONFIG(dev);
+	int tach_ch = config->channel;
+	int p = config->pulses_per_round;
+	struct tach_it8xxx2_data *const data = DRV_DATA(dev);
+
+	if (chan != SENSOR_CHAN_RPM) {
+		LOG_ERR("Sensor chan %d, only support SENSOR_CHAN_RPM", chan);
+		return -ENOTSUP;
+	}
+
+	/* Transform count unit to RPM */
+	if (data->capture > 0) {
+		if ((tach_ch == TACH0_CHANNEL_A) || (tach_ch == TACH0_CHANNEL_B))
+			/*
+			 * Fan Speed (RPM) = 60 / (1/fs * {F1TMRR, F1TLRR} * P)
+			 * - P denotes the numbers of pulses per round
+			 * - {F1TMRR, F1TLRR} = 0000h denotes Fan Speed is zero
+			 * - The sampling rate (fs) is TACH_FREQ / 128
+			 */
+			val->val1 = (60 * TACH_FREQ / 128 / p / (data->capture));
```

Comment:
Can `data->capture` ever be zero?

Reply:
Yes,  1. when fan stop and tach data is valid, data->capture is zero, or
         2. tach data isn't valid, data->capture is zero.
 
But only data->capture > 0, we transform count to R.P.M by this formula, otherwise we set val->val1 = 0 directly.